### ISSUE TITLE
Core Dump HHVM 3.7.1

### ISSUE THREAD

**Issue by brianlmoon** [facebook/hhvm#5497]
This happened during a timed AJAX call from my web browser being left open over night. The same call had happened hundreds of times. There was nothing special about this one.

Trusty package info:
```
Package: hhvm
Status: install ok installed
Priority: optional
Section: web
Maintainer: Paul Tarjan <pt@fb.com>
Architecture: amd64
Version: 3.7.1-1~trusty
Replaces: hhvm-fastcgi
Provides: hhvm-fastcgi
Depends: binutils, libasn1-8-heimdal, libboost-filesystem1.54.0, libboost-program-options1.54.0, libboost-system1.54.0, libboost-system1.54.0, libboost-regex1.54.0, libbz2-1.0, libc6, libcap2, libc-client2007e, libcomerr2, libelf1, libexpat1, libfontconfig1, libfreetype6, libgcc1, libgcrypt11, libgd3, libgnutls26, libgpg-error0, libgssapi3-heimdal, libgssapi-krb5-2, libhcrypto4-heimdal, libheimbase1-heimdal, libheimntlm0-heimdal, libhx509-5-heimdal, libicu52, libidn11, libjpeg-turbo8, libk5crypto3, libkeyutils1, libkrb5-26-heimdal, libkrb5-3, libkrb5support0, libldap-2.4-2, libmcrypt4, libmemcached10, libmysqlclient18, libonig2, libp11-kit0, libpam0g, libpcre3, libpng12-0, libroken18-heimdal, librtmp0, libsasl2-2, libsqlite3-0, libssl1.0.0, libstdc++6, libtasn1-6, libtbb2, libtinfo5, libwind0-heimdal, libx11-6, libxau6, libxcb1, libxdmcp6, libxml2, libxpm4, zlib1g, libedit2, libjemalloc1 (>= 3.0.0), libgoogle-glog0 (>= 0.3.1), libcurl3 (>= 7.28.0), libboost-thread1.54.0, libboost-context1.54.0, libmagickwand5, libxslt1.1, libevent-2.0-5, libgmp10, libyaml-0-2
Conflicts: hhvm-fastcgi
Conffiles:
 /etc/hhvm/server.ini 0b0b3ab7fc11bee3b61271f870d81378
 /etc/hhvm/php.ini 23f945471468111f2b48a539ff9fd241
 /etc/default/hhvm d60fa72d85c37c4457c0e14a3272fa3f
 /etc/init.d/hhvm 211d392430226d8ae94a4d613836e772
 /etc/nginx/hhvm.conf 5c157e50de5444bf970a4ec7fbbf234c
 /etc/apache2/mods-available/hhvm_proxy_fcgi.conf 6ab1639f98500745d1f96cd48e269059
Description: HHVM virtual machine, runtime, and JIT for the PHP language
Description-md5: 98571cd7333493baf95470adfb419c6f
```

```
# uname -a
Linux brianm 3.13.0-53-generic #89-Ubuntu SMP Wed May 20 10:34:39 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```
Error log:
```
[Tue Jun 16 11:05:31 2015] [hphp] [48255:7f6985bff700:1572:000001] [7f69afbf3d40:7f69afbf3cc9:7f69afbf70d8:220f09d:141e80e:133ded9:133f34a:133f5a4:133f93a:2c72014:2c71e8a:440027b:4400001:146c806:132c82b:122e5a9:2c6121f:122ffb5:199a1c9:45a432e:45a19be:46d24cf:45d6bfa:45d840e:45b7d7a:c000000:4400001:146c806:132c82b:122e5a9:2c6121f:2cb0039:2cafebf:fc00b2:dcb142:11091fa:110b2f2:111f05a:111f5ce:111bff6:21f4b67:f8c287:7f69b07aa182:7f69afcb747d] Core dumped: Aborted
[Tue Jun 16 11:05:31 2015] [hphp] [48255:7f6985bff700:1572:000002] [7f69afbf3d40:7f69afbf3cc9:7f69afbf70d8:220f09d:141e80e:133ded9:133f34a:133f5a4:133f93a:2c72014:2c71e8a:440027b:4400001:146c806:132c82b:122e5a9:2c6121f:122ffb5:199a1c9:45a432e:45a19be:46d24cf:45d6bfa:45d840e:45b7d7a:c000000:4400001:146c806:132c82b:122e5a9:2c6121f:2cb0039:2cafebf:fc00b2:dcb142:11091fa:110b2f2:111f05a:111f5ce:111bff6:21f4b67:f8c287:7f69b07aa182:7f69afcb747d] Stack trace in /tmp/stacktrace.48255.log
```
Stack trace:
```
Host: brianm
ProcessID: 48256
ThreadID: 7f6985bff700
ThreadPID: 64298
Name: unknown program
Type: Aborted
Runtime: hhvm
Version: tags/HHVM-3.7.1-0-g111c76378ac29aa8f25e6bb3c5f15b628e6bbea1
DebuggerCount: 0

-------------------------------------IRUnit-------------------------------------
: digraph G {
B0 -> B1
B1 -> B3; B1 -> B2 [color=cyan]
B3 -> B6; B3 -> B4
B6 -> B5
B4 -> B5
B5 -> B11; B5 -> B7 [color=green]
B11 -> B8 [color=cyan]
B7 -> B10 [color=green]; B7 -> B9 [color=blue]
B10 -> B8 [color=cyan]
}

 B0:
Pseudo-main at 0
maxStackCells: 1
numLocals: 0
numIterators: 0
    --- bc 0, fp _, spOff 0 ()
   0: Int 1
    (00) t0:FramePtr = DefFP
    (01) t1:StkPtr = DefSP<0> t0:FramePtr
    (02) Jmp -> B1
  no fallthrough

 B1: (preds B0)
    --- bc 0, fp 0, spOff 0 () [profTrans=-325492]
   0: Int 1
    (03) EndGuards

    --- bc 9, fp 0, spOff 1 () [profTrans=-325492]
   9: RetC
    (08) ReleaseVVOrExit t0:FramePtr -> B2<Unlikely>
  -> B3

 B3: (preds B1)
    (09) GenericRetDecRefs t0:FramePtr
    (10) t4:{Obj|Cctx} = LdCtx t0:FramePtr
    (11) CheckCtxThis t4:{Obj|Cctx} -> B4
  -> B6

 B6: (preds B3)
    (12) t5:Obj = CastCtxThis t4:{Obj|Cctx}
    (13) DecRef t5:Obj
    (14) Jmp -> B5
  no fallthrough

 B4: (preds B3)
    (15) Jmp -> B5
  no fallthrough

 B5: (preds B4 B6)
    --- bc 9, fp 0, spOff 0 () [profTrans=-325492]
   9: RetC
    (16) CheckSurpriseFlags -> B7<Unlikely>
  -> B11

 B11: (preds B5)
    (24) Jmp -> B8
  no fallthrough

 B8: (preds B11 B10)
    (20) StRetVal t0:FramePtr, 1
    (21) t6:StkPtr = RetAdjustStk t0:FramePtr
    (22) RetCtrl<0> t6:StkPtr, t0:FramePtr
  no fallthrough

------------------------cold blocks-------------------------
 B7<Unlikely>: (preds B5)
    (19) ReturnHook t0:FramePtr, 1 -> B9<Catch>
  -> B10<Unlikely>

 B10<Unlikely>: (preds B7)
    (23) Jmp -> B8
  no fallthrough

 B2<Unlikely>: (preds B1)
    --- bc 9, fp 0, spOff 1 () [profTrans=-325492]
   9: RetC
    (06) StStk<IrSpOff -1> t1:StkPtr, 1
    (07) InterpOneCF<RetC: spOff:-1, bcOff:9, popped:0, pushed:0> t1:StkPtr, t0:FramePtr
  no fallthrough

-----------------------frozen blocks------------------------
 B9<Catch>: (preds B7)
    --- bc 9, fp 0, spOff 0 () [profTrans=-325492]
   9: RetC
    (17) BeginCatch
    (18) EndCatch<IrSpOff 0> t0:FramePtr, t1:StkPtr
  no fallthrough

-----------------------------------RegionDesc-----------------------------------
: Region (1 blocks):
Block -325492 @0 length 2 initSpOff 0 inlineLevel 0
    0  Int 1
    9  RetC
Arcs:
Side-exiting Blocks:


ThreadType: Web Request
Assertion Message: data block = cold
message: Attempted to emit 1 byte(s) into a 75 byte DataBlock with 0 bytes available. This almost certainly means the TC is full. If this is the case, increasing Eval.JitASize, Eval.JitAColdSize, Eval.JitAFrozenSize and Eval.JitGlobalDataSize in the configuration file when running this script or application should fix this problem.

Assertion Failure: /tmp/tmp.CvLsK49lmP/hphp/runtime/vm/jit/translate-region.cpp:719: HPHP::jit::TranslateResult HPHP::jit::{anonymous}::mcGenRegion(HPHP::jit::IRGS&, const HPHP::jit::RegionDesc&, HPHP::jit::RegionBlacklist&): assertion `0' failed.
URL: /admin/phews/queue/queue_server.php?cmd=get_inprogress_update
Server_SERVER_NAME: brianm.secure.dev.dealnews.com
Server: brianm.secure.dev.dealnews.com

# 0  gsignal at /lib/x86_64-linux-gnu/libc.so.6:0
# 1  abort at /lib/x86_64-linux-gnu/libc.so.6:0
# 2  HPHP::assert_fail(char const*, char const*, unsigned int, char const*, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) at /usr/bin/hhvm:0
# 3  HPHP::jit::translateRegion(HPHP::jit::IRGS&, HPHP::jit::RegionDesc const&, HPHP::hphp_hash_set<HPHP::jit::ProfSrcKey, HPHP::jit::ProfSrcKey::Hasher, std::equal_to<HPHP::jit::ProfSrcKey> >&, HPHP::jit::TransFlags, std::vector<HPHP::jit::RegionDesc::TypePred, std::allocator<HPHP::jit::RegionDesc::TypePred> >&) at /usr/bin/hhvm:0
# 4  HPHP::jit::MCGenerator::translateWork(HPHP::jit::TranslArgs const&) at /usr/bin/hhvm:0
# 5  HPHP::jit::MCGenerator::translate(HPHP::jit::TranslArgs const&) at /usr/bin/hhvm:0
# 6  HPHP::jit::MCGenerator::retranslate(HPHP::jit::TranslArgs const&) at /usr/bin/hhvm:0
# 7  HPHP::jit::MCGenerator::createTranslation(HPHP::jit::TranslArgs const&) at /usr/bin/hhvm:0
# 8  HPHP::jit::MCGenerator::getCallArrayPrologue(HPHP::Func*) at /usr/bin/hhvm:0
# 9  HPHP::jit::funcBodyHelper(HPHP::ActRec*, void*) at /usr/bin/hhvm:0
# 10 HPHP::jit::x64::(anonymous namespace)::BackEnd::enterTCHelper(unsigned char*, HPHP::ActRec*) at back-end-x64.cpp:0
# 11 HPHP::jit::MCGenerator::enterTC(unsigned char*, HPHP::ActRec*) at /usr/bin/hhvm:0
# 12 HPHP::enterVM(HPHP::ActRec*, HPHP::StackArgsState, HPHP::Resumable*, HPHP::ObjectData*) at bytecode.cpp:0
# 13 HPHP::ExecutionContext::invokeFunc(HPHP::TypedValue*, HPHP::Func const*, HPHP::Variant const&, HPHP::ObjectData*, HPHP::Class*, HPHP::VarEnv*, HPHP::StringData*, HPHP::ExecutionContext::InvokeFlags) at /usr/bin/hhvm:0
# 14 HPHP::ExecutionContext::createFunction(HPHP::String const&, HPHP::String const&) at /usr/bin/hhvm:0
# 15 HPHP::f_create_function(HPHP::String const&, HPHP::String const&) at /usr/bin/hhvm:0
# 16 HPHP::jit::x64::(anonymous namespace)::BackEnd::enterTCHelper(unsigned char*, HPHP::ActRec*) at back-end-x64.cpp:0
# 17 HPHP::jit::MCGenerator::enterTC(unsigned char*, HPHP::ActRec*) at /usr/bin/hhvm:0
# 18 HPHP::enterVM(HPHP::ActRec*, HPHP::StackArgsState, HPHP::Resumable*, HPHP::ObjectData*) at bytecode.cpp:0
# 19 HPHP::ExecutionContext::invokeFunc(HPHP::TypedValue*, HPHP::Func const*, HPHP::Variant const&, HPHP::ObjectData*, HPHP::Class*, HPHP::VarEnv*, HPHP::StringData*, HPHP::ExecutionContext::InvokeFlags) at /usr/bin/hhvm:0
# 20 HPHP::ExecutionContext::invokeUnit(HPHP::TypedValue*, HPHP::Unit const*) at /usr/bin/hhvm:0
# 21 HPHP::invoke_file(HPHP::String const&, bool, char const*) at builtin-functions.cpp:0
# 22 HPHP::include_impl_invoke(HPHP::String const&, bool, char const*) at /usr/bin/hhvm:0
# 23 HPHP::hphp_invoke(HPHP::ExecutionContext*, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, HPHP::Array const&, HPHP::VRefParamValue const&, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool&, std::basic_string<char, std::char_traits<char>, std::allocator<char> >&, bool, bool, bool) at /usr/bin/hhvm:0
# 24 HPHP::HttpRequestHandler::executePHPRequest(HPHP::Transport*, HPHP::RequestURI&, HPHP::SourceRootInfo&, bool) at /usr/bin/hhvm:0
# 25 HPHP::HttpRequestHandler::handleRequest(HPHP::Transport*) at /usr/bin/hhvm:0
# 26 HPHP::ServerWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::FastCGITransportTraits>::doJobImpl(std::shared_ptr<HPHP::FastCGIJob>, bool) at /usr/bin/hhvm:0
# 27 HPHP::ServerWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::FastCGITransportTraits>::doJob(std::shared_ptr<HPHP::FastCGIJob>) at /usr/bin/hhvm:0
# 28 HPHP::JobQueueWorker<std::shared_ptr<HPHP::FastCGIJob>, HPHP::Server*, true, false, HPHP::JobQueueDropVMStack>::start() at /usr/bin/hhvm:0
# 29 HPHP::AsyncFuncImpl::ThreadFunc(void*) at /usr/bin/hhvm:0
# 30 HPHP::start_routine_wrapper(void*) at /usr/bin/hhvm:0
# 31 start_thread at /lib/x86_64-linux-gnu/libpthread.so.0:0
# 32 clone at /lib/x86_64-linux-gnu/libc.so.6:0

PHP Stacktrace:

#0  include() called at [/dealnews/mywww/brianm/lib/ORM/Queue.php:1052]
#1  ORM_Queue::parse_queue_wiki_text([redacted, available upon request]) called at [/dealnews/mywww/brianm/lib/ORM/Queue.php:699]
#2  ORM_Queue->load(970711) called at [/dealnews/mywww/brianm/virthost/secure.dealnews/admin/phews/queue/queue_functions.php:80]
#3  get_inprogress() called at [/dealnews/mywww/brianm/virthost/secure.dealnews/admin/phews/queue/queue_functions.php:112]
#4  get_inprogress_html() called at [/dealnews/mywww/brianm/virthost/secure.dealnews/admin/phews/queue/queue_server.php:242]
#5  get_inprogress_update() called at [/dealnews/mywww/brianm/virthost/secure.dealnews/admin/phews/queue/queue_server.php:328]
#6  include(/dealnews/mywww/brianm/virthost/secure.dealnews/admin/phews/queue/queue_server.php) called at [/dealnews/mywww/brianm/virthost/secure.dealnews/admin/router.php:155]
```

**Comment by brianlmoon** [comment]
Hmm, seems this may be a config issue. If so, please ignore.

**Comment by fredemmott** [comment]
Possibly. f_create_function() is suspicious.

If it's not configuration, this might be fixed by replacing '$foo = create_function(...){"..."}' with '$foo = function(...) { ... }'

**Comment by paulbiss** [comment]
We have some fixes for create_function in 3.7.2 that just went out. This is probably related.

**Comment by paulbiss** [comment]
Please reopen if it continues failing in 3.7.2.
